File: 1020.java
import java.util.Scanner;
import java.util.Arrays;
public class Main{
 static int c[]=new int[11];//c[i]存放边长为i的小正方形的个数
 static int d[]=new int[41];//d[i]表示第i行填入小正方形后的最大横坐标
 static int s,n,sum;
 static boolean  ok;

public static void main(String args[])
{
    int t,it,i,tp;
    Scanner sc=new Scanner(System.in);
    t=sc.nextInt();//测试次娄

    for(it=1;it<=t;it++)//循环进行每一次测试
    {
        s=sc.nextInt();//“盘子”大正方形的边长
        n=sc.nextInt();//小正方形的个数
        Arrays.fill(c,0);//每次测试前清空
         ok=false;
        for(i=1;i<=40;i++) d[i]=1;//
        sum=0;
        for(i=1;i<=n;i++)//从命令行依次读入小正方形的边长
        {
            tp=sc.nextInt();
            c[tp]++;//边长为tp的小正方形的个数增加1
            sum+=tp*tp;//统计面积
        }
        if(sum!=s*s) ok=false;//如果所有小正方形面积之和不等于“盘子”面积
        else {ok=false;dfs(0);}//深度优先搜索

        if(ok) System.out.println("KHOOOOB!");
        else System.out.println("HUTUTU!");
    }
   }

  public static void dfs(int a)
{
    int i,j,put,p=0;
    boolean f;
    if(a==n) {//如果对所有的n个小正方形搜索完毕。
       ok=true;
       return;
    }

    //寻找未填的横坐标最小的点的纵坐标
    for(i=1,put=41;i<=s;i++){
        if(d[i]< put) {
          put=d[i];//put:横坐标
          p=i;//p:纵坐标
       }
    }

     //从横坐标最小的那列开始，从大到小搜索能填入的小正方形
    for(i=10;i>=1;i--)
        if(c[i]>0 && put+i-1<=s && p+i-1<=s)//在坐标(put,p)处能放得下小正方形c[i]
        {
            for(j=p,f=true;j<=p+i-1;j++)
                if(d[j]>put) {f=false;break;}
            if(f)
            {
                for(j=p;j<=p+i-1;j++) d[j]+=i;//标记第i行填入小正方形后的最大横坐标

                c[i]--;
                dfs(a+1);
                c[i]++;//回朔
                for(j=p;j<=p+i-1;j++) d[j]-=i;
            }
        }
    }
}

File: 1020.cc
#include <iostream>
using namespace std;
int cake[11], pos[41];  //题目限制的是蛋糕的最大变长为10,且需分割的最大边长为40
int m, k;

bool DFS(int a) {
  int i, min_pos = 41, lie = 0, j, t;
  if (a == k)  //如果遍历到放入的数目和要求的数目相同即完成目标
    return true;
  for (i = 1; i <= m; i++)
    if (pos[i] < min_pos) {
      min_pos = pos[i];  //找到了在竖直方向上用去的最短行数
      lie = i;
    }
  for (i = 10; i >= 1; i--) {
    if (cake[i] && min_pos + i <= m &&
        lie + i - 1 <= m) {  //可以放入边长为I的正方形
      for (j = lie; j <= lie + i - 1; j++)
        if (pos[j] > min_pos)  //该空隙的宽度大于等于i
          break;
      if (j > lie + i - 1) {
        cake[i]--;  //这个就是典型的DFS算法的了
        for (t = lie; t <= lie + i - 1; t++) pos[t] = pos[t] + i;
        if (DFS(a + 1)) return true;  // I GET THE WA HERE
        cake[i]++;                    //回溯部分的算法
        for (t = lie; t <= lie + i - 1; t++) pos[t] = pos[t] - i;
      }
    }
  }
  return false;
}

int main() {
  int n, sum_cake, ori_cake, i, s;
  scanf("%d", &n);
  while (n--) {
    memset(cake, 0, sizeof(cake));
    memset(pos, 0, sizeof(pos));
    scanf("%d%d", &m, &k);
    sum_cake = 0;
    ori_cake = m * m;
    for (i = 1; i <= k; i++) {
      scanf("%d", &s);
      cake[s]++;
      sum_cake = sum_cake + s * s;
    }
    // printf("sum_cake==%d,ori_cake==%d\n",sum_cake,ori_cake);
    if (sum_cake != ori_cake)
      printf("HUTUTU!\n");
    else if (DFS(0))
      printf("KHOOOOB!\n");
    else
      printf("HUTUTU!\n");
  }
  return 0;
}
